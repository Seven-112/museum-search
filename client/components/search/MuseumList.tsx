import gql from "graphql-tag";
import { graphql } from "react-apollo";

/** MuseumList component props. */
interface MuseumListProps {
  query?: string;
}

/** MuseumList query response. */
interface MuseumListResponse {
  museums: any;
}

/** MuseumList query variables. */
interface MuseumListVariables {
  query?: string;
}

/** MuseumList query. */
const GET_MUSEUM_LIST = gql`
  query($query: String) {
    museums(query: $query, first: 50) {
      edges {
        node {
          id
          name
          streetAddress
          city
          state
          latitude
          longitude
        }
      }
      count
    }
  }
`;

/** Higher-order component to provide data from the GraphQL API to a React component. */
const withMuseumList = graphql<
  MuseumListProps,
  MuseumListResponse,
  MuseumListVariables
>(GET_MUSEUM_LIST, {
  options: ({ query }) => ({
    variables: {
      query
    }
  })
});

/** MuseumList component. */
export const MuseumList = withMuseumList(function MuseumList({
  data: { loading, error, museums }
}) {
  return (
    <ul className="list-group">
      {loading ? (
        "Loading..."
      ) : error ? (
        <div className="alert alert-danger">{error}</div>
      ) : museums ? (
        museums.edges.map(edge => (
          <li key={edge.node.id} className="list-group-item">
            <strong>{edge.node.name}</strong>
            <div>{edge.node.streetAddress}</div>
            <div>
              {[edge.node.city, edge.node.state]
                .filter(it => it || false)
                .join(" - ")}
            </div>
          </li>
        ))
      ) : null}
    </ul>
  );
});
