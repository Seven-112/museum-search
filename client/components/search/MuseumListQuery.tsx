import gql from "graphql-tag";
import { Query, QueryProps } from "react-apollo";

type MuseumListQueryProps = Partial<QueryProps<any, any>>;

const GET_MUSEUM_LIST = gql`
  query($query: String! = "museum") {
    museums(query: $query, first: 200) {
      edges {
        node {
          id
          name
          streetAddress
          city
          state
          latitude
          longitude
        }
      }
      count
    }
  }
`;

export const MuseumListQuery: React.StatelessComponent<
  MuseumListQueryProps
> = props => (
  <Query query={GET_MUSEUM_LIST} {...props}>
    {result => props.children(result)}
  </Query>
);
