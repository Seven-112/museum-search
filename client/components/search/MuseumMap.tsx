import gql from "graphql-tag";
import { divIcon, icon, LatLngBounds, point } from "leaflet";
import "leaflet/dist/leaflet.css";
import React from "react";
import { graphql } from "react-apollo";
import { Map, Marker, TileLayer } from "react-leaflet";

/** Callback function called when the map is moved. */
export type MoveHandler = (box: LatLngBounds) => void;

/** MuseumMap component props. */
export interface MuseumMapProps {
  query?: string;
  boundingBox?: object;
  onMove?: MoveHandler;
}

/** MuseumMap query response. */
interface MuseumMapResponse {
  museumMapObjects?: any;
}

/** MuseumMap query variables. */
interface MuseumMapVariables {
  query?: string;
  boundingBox?: object;
}

/** MuseumMapObjects query. */
export const GET_MUSEUM_MAP_OBJECTS = gql`
  query museumMapObjects($query: String, $boundingBox: GeoBoundingBoxInput) {
    museumMapObjects(query: $query, boundingBox: $boundingBox) {
      edges {
        ... on MuseumSearchEdge {
          node {
            __typename
            id
            name
            latitude
            longitude
          }
        }
        ... on GeoPointBucketEdge {
          node {
            __typename
            geoHashKey
            latitude
            longitude
            count
          }
        }
      }
    }
  }
`;

/** Higher-order component to provide data from the GraphQL API to a React component. */
const withMuseumMapObjects = graphql<
  MuseumMapProps,
  MuseumMapResponse,
  MuseumMapVariables
>(GET_MUSEUM_MAP_OBJECTS, {
  options: ({ boundingBox, query }) => ({
    variables: {
      boundingBox,
      query
    }
  })
});

/** Marker icon for a single museum on the map. */
const MUSEUM_MARKER_ICON = icon({
  iconUrl: require("leaflet/dist/images/marker-icon.png"),
  shadowUrl: require("leaflet/dist/images/marker-shadow.png"),
  iconAnchor: [13, 40],
  popupAnchor: [0, -30]
});

/** Marker icon for a cluster of museums. */
function bucketMarkerIcon(count: number): any {
  const size = count < 100 ? "small" : count < 1000 ? "medium" : "large";

  return divIcon({
    html: `<div><span>${count}</span></div>`,
    className: `marker-cluster marker-cluster-${size}`,
    iconSize: point(40, 40)
  });
}

/** MuseumMap component. */
export const MuseumMap = withMuseumMapObjects(function MuseumMap({
  data: { museumMapObjects },
  onMove
}) {
  return (
    <Map
      onmove={onMove && (e => onMove(e.target.getBounds()))}
      center={[38.810338, -98.323266]}
      zoom={4}
      className="h-100"
    >
      <TileLayer
        attribution='&amp;copy <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
        noWrap={true}
        url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
      />
      {museumMapObjects &&
        museumMapObjects.edges.map(edge =>
          edge.__typename == "GeoPointBucketEdge" ? (
            <Marker
              key={`bucket_${edge.node.geoHashKey}`}
              position={[edge.node.latitude, edge.node.longitude]}
              icon={bucketMarkerIcon(edge.node.count)}
            />
          ) : (
            <Marker
              key={`museum_${edge.node.id}`}
              position={[edge.node.latitude, edge.node.longitude]}
              icon={MUSEUM_MARKER_ICON}
            />
          )
        )}
    </Map>
  );
});
