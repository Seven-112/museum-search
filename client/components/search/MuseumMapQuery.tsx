import gql from "graphql-tag";
import { Query, QueryProps } from "react-apollo";

type MuseumMapQueryProps = Partial<QueryProps<any, any>>;

const GET_MUSEUM_MAP_OBJECTS = gql`
  query museumMapObjects($query: String!, $boundingBox: GeoBoundingBoxInput) {
    museumMapObjects(query: $query, boundingBox: $boundingBox) {
      edges {
        ... on MuseumSearchEdge {
          node {
            __typename
            id
            name
            latitude
            longitude
          }
        }
        ... on GeoPointBucketEdge {
          node {
            __typename
            geoHashKey
            latitude
            longitude
            count
          }
        }
      }
    }
  }
`;

export const MuseumMapQuery: React.FunctionComponent<
  MuseumMapQueryProps
> = props => (
  <Query query={GET_MUSEUM_MAP_OBJECTS} {...props}>
    {result => props.children(result)}
  </Query>
);
