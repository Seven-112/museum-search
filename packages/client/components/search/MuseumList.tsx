import gql from "graphql-tag";
import { graphql } from "react-apollo";

/** MuseumList component props. */
interface IMuseumListProps {
  query?: string;
}

/** MuseumList query response. */
interface IMuseumListResponse {
  museums: any;
}

/** MuseumList query variables. */
interface IMuseumListVariables {
  query?: string;
}

/** MuseumList query. */
export const GET_MUSEUM_LIST = gql`
  query($query: String) {
    museums(query: $query, first: 50) {
      edges {
        node {
          id
          name
          streetAddress
          city
          state
          latitude
          longitude
        }
      }
      count
    }
  }
`;

/** Higher-order component to provide data from the GraphQL API to a React component. */
const withMuseumList = graphql<
  IMuseumListProps,
  IMuseumListResponse,
  IMuseumListVariables
>(GET_MUSEUM_LIST, {
  options: ({ query }) => ({
    variables: {
      query
    }
  })
});

/** MuseumList component. */
export const MuseumList = withMuseumList(function MuseumListInternal({
  data: { loading, error, museums }
}) {
  return (
    <ul className="list-group">
      {loading && "Loading..."}
      {error && <div className="alert alert-danger">{error.message}</div>}
      {museums &&
        museums.edges.map(edge => (
          <li key={edge.node.id} className="list-group-item">
            <strong>{edge.node.name}</strong>
            <div>{edge.node.streetAddress}</div>
            <div>
              {[edge.node.city, edge.node.state]
                .filter(it => it || false)
                .join(" - ")}
            </div>
          </li>
        ))}
    </ul>
  );
});
