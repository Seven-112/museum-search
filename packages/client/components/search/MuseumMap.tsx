import gql from "graphql-tag";
import { divIcon, LatLngExpression, LatLngTuple, point } from "leaflet";
import "leaflet/dist/leaflet.css";
import { clamp } from "lodash";
import React, { Ref } from "react";
import { graphql } from "react-apollo";
import { Map, Marker, TileLayer } from "react-leaflet";
import "../../style/MuseumMap.css";
import { LoadingSpinner } from "../loading-spinner/LoadingSpinner";

export type MoveHandler = (box: IBoundingBox) => void;

/** MuseumMap component props. */
export interface IMuseumMapProps {
  initialCenter: LatLngExpression;
  query?: string;
  boundingBox?: object;
  leafletMapRef?: Ref<Map>;
  onMove?: MoveHandler;
  highlightedMuseum?: any;
}

/** MuseumMap query response. */
interface IMuseumMapResponse {
  museumMapObjects?: any;
}

/** MuseumMap query variables. */
interface IMuseumMapVariables {
  query?: string;
  boundingBox?: object;
}

export interface IBoundingBox {
  bottom: number;
  left: number;
  right: number;
  top: number;
}

// tslint:disable-next-line: no-var-requires
const BLUE_MARKER_IMG = require("leaflet/dist/images/marker-icon.png");
// tslint:disable-next-line: no-var-requires
const RED_MARKER_IMG = require("../../img/marker-icon-red.png");

/** MuseumMapObjects query. */
export const GET_MUSEUM_MAP_OBJECTS = gql`
  query museumMapObjects($query: String, $boundingBox: GeoBoundingBoxInput) {
    museumMapObjects(query: $query, boundingBox: $boundingBox) {
      edges {
        ... on MuseumSearchEdge {
          node {
            __typename
            id
            name
            latitude
            longitude
          }
        }
        ... on GeoPointBucketEdge {
          node {
            __typename
            geoHashKey
            latitude
            longitude
            count
          }
        }
      }
    }
  }
`;

/** Higher-order component to provide data from the GraphQL API to a React component. */
const withMuseumMapObjects = graphql<
  IMuseumMapProps,
  IMuseumMapResponse,
  IMuseumMapVariables
>(GET_MUSEUM_MAP_OBJECTS, {
  options: ({ boundingBox, query }) => ({
    variables: {
      boundingBox,
      query
    }
  })
});

function museumMarkerIcon(museum: any, img: any) {
  return divIcon({
    className: "museum-marker",
    html: `
      <div style="width: 300px">
        <img src="${img}">
        <div><label>${museum.name}</label></div>
      </div>
    `,
    iconAnchor: [13, 40],
    popupAnchor: [0, -30]
  });
}

/** Marker icon for a cluster of museums. */
function bucketMarkerIcon(count: number): any {
  const size = count < 100 ? "small" : count < 1000 ? "medium" : "large";

  return divIcon({
    className: `marker-cluster marker-cluster-${size}`,
    html: `<div><span>${count}</span></div>`,
    iconSize: point(40, 40)
  });
}

function redHighlightMarker(museum?: any) {
  if (museum) {
    const position: LatLngTuple = [museum.latitude, museum.longitude];

    return (
      <Marker
        position={position}
        icon={museumMarkerIcon(museum, RED_MARKER_IMG)}
      />
    );
  }
  return null;
}

/** MuseumMap component. */
export const MuseumMap = withMuseumMapObjects(function MuseumMapInternal({
  data: { loading, museumMapObjects },
  highlightedMuseum,
  initialCenter,
  leafletMapRef,
  onMove
}) {
  const highlightMarker = redHighlightMarker(highlightedMuseum);

  function onMoveInternal(event) {
    if (!onMove) {
      return;
    }

    const leafletBox = event.target.getBounds();

    const boundingBox: IBoundingBox = {
      bottom: clamp(leafletBox.getSouth(), -90, 90),
      left: clamp(leafletBox.getWest(), -180, 180),
      right: clamp(leafletBox.getEast(), -180, 180),
      top: clamp(leafletBox.getNorth(), -90, 90)
    };

    onMove(boundingBox);
  }

  return (
    <div className="h-100">
      <div style={{ zIndex: 500 }} className="position-absolute w-100">
        <LoadingSpinner loading={loading} />
      </div>
      <Map
        center={initialCenter}
        className="h-100"
        onmove={onMoveInternal}
        ref={leafletMapRef}
        zoom={4}
      >
        <TileLayer
          attribution='&amp;copy <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
          noWrap={true}
          url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
        />
        <MarkerGroup museumMapObjects={museumMapObjects} />
        {highlightMarker}
      </Map>
    </div>
  );
});

/**
 * Marker group component that only re-renders when the passed list of museums changes.
 */
class MarkerGroup extends React.Component<IMuseumMapResponse> {
  public render() {
    const { museumMapObjects } = this.props;

    return (
      <div>
        {museumMapObjects &&
          museumMapObjects.edges.map(edge => {
            switch (edge.__typename) {
              case "GeoPointBucketEdge":
                return (
                  <Marker
                    key={`bucket_${edge.node.geoHashKey}`}
                    position={[edge.node.latitude, edge.node.longitude]}
                    icon={bucketMarkerIcon(edge.node.count)}
                  />
                );
              case "MuseumSearchEdge":
                return (
                  <Marker
                    key={`museum_${edge.node.id}`}
                    position={[edge.node.latitude, edge.node.longitude]}
                    icon={museumMarkerIcon(edge.node, BLUE_MARKER_IMG)}
                  />
                );
            }
          })}
      </div>
    );
  }

  public shouldComponentUpdate({ museumMapObjects }: IMuseumMapResponse) {
    // Only render if the museum list is different.
    return museumMapObjects !== this.props.museumMapObjects;
  }
}
