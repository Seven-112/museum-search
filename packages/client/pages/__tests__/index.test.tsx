import { shallow } from "enzyme";
import { Formik } from "formik";
import { MuseumList } from "../../components/search/MuseumList";
import { MuseumSearchPage } from "../index";

describe("MuseumSearchPage component", () => {
  it("Shallow renders.", () => {
    const mockRouter = {
      query: {}
    };
    const wrapper = shallow(<MuseumSearchPage router={mockRouter as any} />);
    expect(wrapper).toMatchSnapshot("MuseumSearchPage shallow render.");
  });

  it("Performs a search using the search input.", () => {
    const mockRouterPush = jest.fn();

    const mockRouter = {
      push: mockRouterPush,
      query: {}
    };
    const wrapper = shallow(<MuseumSearchPage router={mockRouter as any} />);

    // Submit the search form.
    (wrapper.find(Formik).prop("onSubmit") as any)({ query: "space museum" });

    expect(mockRouterPush).toHaveBeenCalledTimes(1);
    expect(mockRouterPush).lastCalledWith({
      pathname: "/",
      query: { q: "space museum" }
    });
  });

  it("Performs a bounding-box search when the map is moved.", async () => {
    // Un-debounce the onMapMove function for this test.
    jest.spyOn(require("lodash"), "debounce").mockImplementationOnce(fn => fn);

    const mockRouter = {
      query: {}
    };
    const wrapper = shallow(<MuseumSearchPage router={mockRouter as any} />);

    // The MuseumMap loaded dynamically, so it is selected as "LoadableComponent".
    (wrapper.find("LoadableComponent").prop("onMove") as any)({
      target: {
        getBounds() {
          return {
            getEast: () => 4,
            getNorth: () => 1,
            getSouth: () => 3,
            getWest: () => 2
          };
        }
      }
    });

    expect(wrapper.find("LoadableComponent").prop("boundingBox")).toEqual({
      bottom: 3,
      left: 2,
      right: 4,
      top: 1
    });
  });

  it("Passes the search query into child components.", async () => {
    const mockRouter = {
      query: { q: "zoo" }
    };
    const wrapper = shallow(<MuseumSearchPage router={mockRouter as any} />);

    expect(wrapper.find(MuseumList).prop("query")).toEqual("zoo");

    // The MuseumMap loaded dynamically, so it is selected as "LoadableComponent".
    expect(wrapper.find("LoadableComponent").prop("query")).toEqual("zoo");
  });
});
