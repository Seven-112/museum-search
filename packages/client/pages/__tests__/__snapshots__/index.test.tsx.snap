// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MuseumSearchPage component Shallow renders.: MuseumSearchPage shallow render. 1`] = `
<div
  className="container-fluid d-flex flex-column flex-fill"
>
  <style>
    
  html, body {
    margin: 0;
    height: 100%
  }
  #__next {
    display: flex;
    flex-flow: column;
    height: 100%;
  }

  </style>
  <Head
    title="Museum Search"
  />
  <div
    className="flex-shrink-0"
  >
    <h1
      className="col-12"
    >
      Museum Search
    </h1>
  </div>
  <div
    className="row flex-fill"
  >
    <div
      className="col-md-3 p-0"
    >
      <div
        className="card h-100"
      >
        <div
          className="card-body flex-grow-0"
        >
          <Formik
            enableReinitialize={false}
            initialValues={
              Object {
                "query": undefined,
              }
            }
            isInitialValid={false}
            onSubmit={[Function]}
            validateOnBlur={true}
            validateOnChange={true}
          >
            <Form>
              <FormikConnect(FieldInner)
                autoComplete="off"
                className="form-control"
                name="query"
                placeholder="Search..."
              />
            </Form>
          </Formik>
        </div>
        <div
          style={
            Object {
              "overflowY": "scroll",
            }
          }
        >
          <Apollo(MuseumList)
            query="museum"
          />
        </div>
      </div>
    </div>
    <div
      className="col-md-9 p-0"
    >
      <LoadableComponent
        onMove={[Function]}
      />
    </div>
  </div>
</div>
`;
